Engine/Commands/InvokeScriptAnalyzerCommand.cs

```diff
@@ -167,6 +167,18 @@ public SwitchParameter SuppressedOnly
         }
         private bool suppressedOnly;
 
+        /// <summary>
+        /// ShowAll: Show the suppressed and non-suppressed message
+        /// </summary>
+        [Parameter(Mandatory = false)]
```

Comment:
This isn't mutually exclusive with `SuppressedOnly` and probably should be. What should a user expect when both are provided?

Given that `-SuppressedOnly` already exists and we don't want to remove or otherwise break it, we probably can't move to an enum parameter like `-SuppressionPreference <Exclude | Include | SuppressedOnly>`.

An alternative might be to put each switch into its own parameter set, but that starts to get complicated. We should discuss this further.

Reply:
```suggestion
        [Parameter]
```

`Mandatory` defaults to `false`

Reply:
correct, if it's not mandatory, that should just be omitted. if `SuppressedOnly` should not be used with this, then parameter sets should be used (and it does get complicated). Although it doesn't look to complicated at the moment it would mean an additional 2 parameter sets.

Reply:
yeah. I totally agree with you guys on that point.
If you decided to change the parameter set, let's discuss more details on the design, I'd like to help.

Reply:
From our offline conversation:

> If the parameter set change is too complicated, can we check in the current solution firstly?

Yeah I think it's a good idea for us to separate out the suppression combining parts from the new parameter/feature. We can then discuss offline or over a call what the best way forward is.

Reply:
To maintain PSSA's existing design, I think we can make '-SuppressedOnly' / '-IncludeSuppressions' exclusive in the short term. What do you think about it? Look forward to your reply.

Reply:
Btw,  `ExcludeRule` and `IncludeRule`  should also be mutually exclusive.

Reply:
> make '-SuppressedOnly' / '-IncludeSuppressions' exclusive

Yeah, they should be in different parameter sets.

> Btw, ExcludeRule and IncludeRule should also be mutually exclusive.

No, those can be used together to include a non-default rule while excluding a default one.

Reply:
To make `-SuppressedOnly` / `-IncludeSuppressions` exclusive.
I used the logic below:
```
 Function TestParamSet {
    [CmdletBinding()]
    Param
    (
	[Parameter(Position = 0,ParameterSetName = "File1",Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = "File2,Mandatory = true"]
        [switch] $Path,
        
	[Parameter(Position = 0,ParameterSetName = "ScriptDefinition1,Mandatory = true")]
        [Parameter(Position = 0,ParameterSetName = "ScriptDefinition2,Mandatory = true")]
	[switch] $ScriptDefinition,
        
        
        [Parameter(ParameterSetName = 'File1', Mandatory = $false)]
        [Parameter(ParameterSetName = 'ScriptDefinition1', Mandatory = $false)]
        [switch] $SuppressedOnly,

        [Parameter(ParameterSetName = 'File2', Mandatory = $false)]
        [Parameter(ParameterSetName = 'ScriptDefinition2', Mandatory = $false)]
        [switch] $IncludeSuppressions
    )

    Process {
        #Do Nothing
    }
}
```
So we've decided on this solution, right?
LMK if you have better suggestions. Thanks

Reply:
To make -SuppressedOnly / -IncludeSuppressions exclusive.
I used the logic below:

 Function TestParamSet {
    [CmdletBinding()]
    Param
    (
	[Parameter(Position = 0,ParameterSetName = "File1",Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = "File2",Mandatory = true)]
        [switch] $Path,
        
	[Parameter(Position = 0,ParameterSetName = "ScriptDefinition1",Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = "ScriptDefinition2",Mandatory = true)]
	[switch] $ScriptDefinition,
        
        
        [Parameter(ParameterSetName = "File1", Mandatory = $false)]
        [Parameter(ParameterSetName = "ScriptDefinition1', Mandatory = $false)]
        [switch] $SuppressedOnly,

        [Parameter(ParameterSetName = "File2", Mandatory = $false)]
        [Parameter(ParameterSetName = "ScriptDefinition2", Mandatory = $false)]
        [switch] $IncludeSuppressions
    )

    Process {
        #Do Nothing
    }
}
```
So we've decided on this solution, right?
LMK if you have better suggestions. Thanks

Reply:
To make -SuppressedOnly / -IncludeSuppressions exclusive.
I used the logic below:
```
 Function TestParamSet {
    [CmdletBinding()]
    Param
    (
	[Parameter(Position = 0,ParameterSetName = "File1",Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = "File2,Mandatory = true"]
        [switch] $Path,
        
	[Parameter(Position = 0,ParameterSetName = "ScriptDefinition1,Mandatory = true")]
        [Parameter(Position = 0,ParameterSetName = "ScriptDefinition2,Mandatory = true")]
	[switch] $ScriptDefinition,
        
        
        [Parameter(ParameterSetName = 'File1', Mandatory = $false)]
        [Parameter(ParameterSetName = 'ScriptDefinition1', Mandatory = $false)]
        [switch] $SuppressedOnly,

        [Parameter(ParameterSetName = 'File2', Mandatory = $false)]
        [Parameter(ParameterSetName = 'ScriptDefinition2', Mandatory = $false)]
        [switch] $IncludeSuppressions
    )

    Process {
        #Do Nothing
    }
}
```
So we've decided on this solution, right?
LMK if you have better suggestions. Thanks

Reply:
To make -SuppressedOnly / -IncludeSuppressions exclusive.
I used the logic below:
```
 Function TestParamSet {
    [CmdletBinding()]
    Param
    (
	[Parameter(Position = 0,ParameterSetName = 'File1',Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = 'File2',Mandatory = true]
        [switch] $Path,
        
	[Parameter(Position = 0,ParameterSetName = 'ScriptDefinition1',Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = 'ScriptDefinition2',Mandatory = true)]
	[switch] $ScriptDefinition,
        
        
        [Parameter(ParameterSetName = 'File1', Mandatory = false)]
        [Parameter(ParameterSetName = 'ScriptDefinition1', Mandatory = false)]
        [switch] $SuppressedOnly,

        [Parameter(ParameterSetName = 'File2', Mandatory = false)]
        [Parameter(ParameterSetName = 'ScriptDefinition2', Mandatory = false)]
        [switch] $IncludeSuppressions

        [Parameter(Position = 0,ParameterSetName = 'File1',Mandatory = false)]
        [Parameter(Position = 0,ParameterSetName = 'File2',Mandatory = false]
        [switch] $Fix,
    )

    Process {
        #Do Nothing
    }
}
```
So we've decided on this solution, right?
I set the cmdlet `DefaultParameterSetName = "File1"`
so it's to say:
1、`Invoke-ScriptAnalyzer C:\build.ps1`  --> ParameterSetName==DefaultParameterSetName .

2、`Invoke-ScriptAnalyzer -Path C:\build.ps1` -SuppressedOnly`  --> ParameterSetName=="File1".

3、`Invoke-ScriptAnalyzer -Path C:\build.ps1` -IncludeSuppressions`  --> ParameterSetName=="File2".

4、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp`  --> **set ParameterSetName="ScriptDefinition1" first.**

5、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp -SuppressedOnly`  --> ParameterSetName=="ScriptDefinition1" 

6、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp -IncludeSuppressions`  -->ParameterSetName=="ScriptDefinition2" 

LMK if you have any suggestions, Thanks!

Reply:
To make -SuppressedOnly / -IncludeSuppressions exclusive.
I used the logic below:
```
 Function TestParamSet {
    [CmdletBinding()]
    Param
    (
	[Parameter(Position = 0,ParameterSetName = 'File1',Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = 'File2',Mandatory = true]
        [switch] $Path,
        
	[Parameter(Position = 0,ParameterSetName = 'ScriptDefinition1',Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = 'ScriptDefinition2',Mandatory = true)]
	[switch] $ScriptDefinition,
        
        
        [Parameter(ParameterSetName = 'File1', Mandatory = false)]
        [Parameter(ParameterSetName = 'ScriptDefinition1', Mandatory = false)]
        [switch] $SuppressedOnly,

        [Parameter(ParameterSetName = 'File2', Mandatory = false)]
        [Parameter(ParameterSetName = 'ScriptDefinition2', Mandatory = false)]
        [switch] $IncludeSuppressions

        [Parameter(Position = 0,ParameterSetName = 'File1',Mandatory = false)]
        [Parameter(Position = 0,ParameterSetName = 'File2',Mandatory = false]
        [switch] $Fix,
    )

    Process {
        #Do Nothing
    }
}
```
So we've decided on this solution, right?
I set the cmdlet DefaultParameterSetName = "File1"
so it's to say:
1、`Invoke-ScriptAnalyzer C:\build.ps1` --> ParameterSetName==DefaultParameterSetName .

2、`Invoke-ScriptAnalyzer -Path C:\build.ps1 -SuppressedOnly` --> ParameterSetName=="File1".

3、`Invoke-ScriptAnalyzer -Path C:\build.ps1 -IncludeSuppressions` --> ParameterSetName=="File2".

4、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp` --> **set ParameterSetName="ScriptDefinition1" first.**

5、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp -SuppressedOnly` --> ParameterSetName=="ScriptDefinition1"

6、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp -IncludeSuppressions` -->ParameterSetName=="ScriptDefinition2"

LMK if you have any suggestions, Thanks!

Reply:
To make -SuppressedOnly / -IncludeSuppressions exclusive.
I used the logic below:
```
 Function TestParamSet {
    [CmdletBinding()]
    Param
    (
	[Parameter(Position = 0,ParameterSetName = 'File1',Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = 'File2',Mandatory = true]
        [switch] $Path,
        
	[Parameter(Position = 0,ParameterSetName = 'ScriptDefinition1',Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = 'ScriptDefinition2',Mandatory = true)]
	[switch] $ScriptDefinition,
        
        
        [Parameter(ParameterSetName = 'File1', Mandatory = false)]
        [Parameter(ParameterSetName = 'ScriptDefinition1', Mandatory = false)]
        [switch] $SuppressedOnly,

        [Parameter(ParameterSetName = 'File2', Mandatory = true)]
        [Parameter(ParameterSetName = 'ScriptDefinition2', Mandatory = true)]
        [switch] $IncludeSuppressions

        [Parameter(Position = 0,ParameterSetName = 'File1',Mandatory = false)]
        [Parameter(Position = 0,ParameterSetName = 'File2',Mandatory = false]
        [switch] $Fix,
    )

    Process {
        #Do Nothing
    }
}
```
So we've decided on this solution, right?
I set the cmdlet DefaultParameterSetName = "File1"
so it's to say:
1、`Invoke-ScriptAnalyzer C:\build.ps1` --> ParameterSetName==DefaultParameterSetName .

2、`Invoke-ScriptAnalyzer -Path C:\build.ps1 -SuppressedOnly` --> ParameterSetName=="File1".

3、`Invoke-ScriptAnalyzer -Path C:\build.ps1-IncludeSuppressions` --> ParameterSetName=="File2".

4、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp` --> ParameterSetName="ScriptDefinition1".

5、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp -SuppressedOnly` --> ParameterSetName=="ScriptDefinition1"

6、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp -IncludeSuppressions `-->ParameterSetName=="ScriptDefinition2"

LMK if you have any suggestions, Thanks!

Reply:
To make -SuppressedOnly / -IncludeSuppressions exclusive.
I used the logic below:
```
 Function TestParamSet {
    [CmdletBinding()]
    Param
    (
	[Parameter(Position = 0,ParameterSetName = 'File1',Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = 'File2',Mandatory = true]
        [switch] $Path,
        
	[Parameter(Position = 0,ParameterSetName = 'ScriptDefinition1',Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = 'ScriptDefinition2',Mandatory = true)]
	[switch] $ScriptDefinition,
        
        
        [Parameter(ParameterSetName = 'File1', Mandatory = false)]
        [Parameter(ParameterSetName = 'ScriptDefinition1', Mandatory = false)]
        [switch] $SuppressedOnly,

        [Parameter(ParameterSetName = 'File2', Mandatory = true)]
        [Parameter(ParameterSetName = 'ScriptDefinition2', Mandatory = true)]
        [switch] $IncludeSuppressions

        [Parameter(Position = 0,ParameterSetName = 'File1',Mandatory = false)]
        [Parameter(Position = 0,ParameterSetName = 'File2',Mandatory = false]
        [switch] $Fix,
    )

    Process {
        #Do Nothing
    }
}
```
So we've decided on this solution, right?
Don't set the cmdlet DefaultParameterSetName
so it's to say:
1、`Invoke-ScriptAnalyzer -Path C:\build.ps1` --> ParameterSetName=="File1".

2、`Invoke-ScriptAnalyzer -Path C:\build.ps1 -SuppressedOnly` --> ParameterSetName=="File1".

3、`Invoke-ScriptAnalyzer -Path C:\build.ps1 -IncludeSuppressions` --> ParameterSetName=="File2".

4、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp` --> ParameterSetName="ScriptDefinition1".

5、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp -SuppressedOnly` --> ParameterSetName=="ScriptDefinition1"

6、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp -IncludeSuppressions` -->ParameterSetName=="ScriptDefinition2"

LMK if you have any suggestions, Thanks!

Reply:
To make -SuppressedOnly / -IncludeSuppressions exclusive.
I used the logic below:
```
 Function TestParamSet {
    [CmdletBinding(DefaultParameterSetName="File1")]
    Param
    (
	[Parameter(Position = 0,ParameterSetName = 'File1',Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = 'File2',Mandatory = true]
        [switch] $Path,
        
	[Parameter(Position = 0,ParameterSetName = 'ScriptDefinition1',Mandatory = true)]
        [Parameter(Position = 0,ParameterSetName = 'ScriptDefinition2',Mandatory = true)]
	[switch] $ScriptDefinition,
        
        
        [Parameter(ParameterSetName = 'File1', Mandatory = false)]
        [Parameter(ParameterSetName = 'ScriptDefinition1', Mandatory = false)]
        [switch] $SuppressedOnly,

        [Parameter(ParameterSetName = 'File2', Mandatory = true)]
        [Parameter(ParameterSetName = 'ScriptDefinition2', Mandatory = true)]
        [switch] $IncludeSuppressions

        [Parameter(Position = 0,ParameterSetName = 'File1',Mandatory = false)]
        [Parameter(Position = 0,ParameterSetName = 'File2',Mandatory = false]
        [switch] $Fix,
    )

    Process {
        #Do Nothing
    }
}
```
so it's to say:

1、`Invoke-ScriptAnalyzer C:\build.ps1` --> ParameterSetName==DefaultParameterSetName.

2、`Invoke-ScriptAnalyzer -Path C:\build.ps1` --> ParameterSetName==DefaultParameterSetName.

3、`Invoke-ScriptAnalyzer -Path C:\build.ps1 -SuppressedOnly` --> ParameterSetName=="File1".

4、`Invoke-ScriptAnalyzer -Path C:\build.ps1 -IncludeSuppressions` --> ParameterSetName=="File2".

5、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp` --> ParameterSetName=="ScriptDefinition1".

6、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp -SuppressedOnly` --> ParameterSetName=="ScriptDefinition1"

7、`Invoke-ScriptAnalyzer -ScriptDefinition $tmp -IncludeSuppressions` -->ParameterSetName=="ScriptDefinition2"

LMK if you have any suggestions, Thanks!